Bluetooth: Expose APIs that convert handles to host objects

rugeGerritsen opened the pull request:
When implementing vendor specific HCI APIs and events, we want to be able to convert between host objects, handles and back again. Currently, the translation from handle to object is unavailable to application developers. 

This PR suggests to expose functions like these.

The current PR only exposes the existing internal APIs. We should probably rename them:

To be decided: 
* Naming of them. Internally we sometimes use `lookup_handle()`, in other places `get_<obj_type>()`. 
  We already have some public APIs using `lookup`, but that
* Where to put the new APIs and their prefix. Should they be added to hci.h or not where the translation object to handle is found?

Suggested naming:
* bt_hci_adv_lookup_handle()`
* bt_hci_per_adv_sync_lookup_handle()`
* bt_hci_conn_lookup_handle()`

rugeGerritsen committed:
Bluetooth: HCI: Expose bt_hci_conn_lookup_handle()

When implementing vendor specific HCI APIs and events,
we want to be able to convert between host objects,
handles and back again.

Exposing this API makes that possible

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

Thalley commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -113,6 +113,16 @@ int bt_hci_cmd_send_sync(uint16_t opcode, struct net_buf *buf,
+ * @param handle The connection handle
+ * @param type   The connection handle type
+ *
+ * @retval The corresponding connection object on success,
+ *         NULL if it does not exist.
```
```suggestion
 * @retval The corresponding connection object on success
 * @retval NULL if it does not exist
```

Isn't it this the proper way of using `@retval`? One line for each return value? Or does the current version also work?

alwa-nordic replied:
`@retval` renders as a table. It's most useful for enumerating values or ranges. It will look incorrect, like this:

> Return values:
> <table>
> <tr><td><code>The<td>corresponding connection object on success,
> <tr><td><code>NULL<td>if it does not exist
> </table>

The alternative is `@returns`:

```suggestion
 * @returns The corresponding connection object on success,
 *         NULL if it does not exist.
```

Or to add in a placeholder like this:

```suggestion
 * @retval NULL if it does not exist.
 * @retval other The corresponding connection object on success,
```

Thalley commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -113,6 +113,16 @@ int bt_hci_cmd_send_sync(uint16_t opcode, struct net_buf *buf,
+/** @brief Get connection given a connection handle.
+ *
+ * @param handle The connection handle
+ * @param type   The connection handle type
```
Where does the user get the `type` from? `bt_hci_get_conn_handle` does not return a type

rugeGerritsen replied:
I would assume that the user knows the handle type based upon context. For example if there is an event that is specific to ACLs or specific to ISO, the user would know what type to use

Thalley replied:
There is of course also just `BT_CONN_TYPE_ALL` which can be used, so this is probably a non-issue

rugeGerritsen replied:
Yes. Some handle types are not overlapping. See Core_v5.4, Vol 4, Part E, Section 5.3.

We could simply use CONN_TYPE_ALL and not let the user specify the handle type. I would guess that this is simpler

Thalley commented on tests/bsim/bluetooth/host/adv/extended/src/ext_adv_advertiser.c:
```diff
@@ -168,12 +169,23 @@ static struct bt_conn_cb conn_cbs = {
 	create_ext_adv_set(&ext_adv, false);
 	start_ext_adv_set(ext_adv);
 
+	err = bt_hci_get_adv_handle(ext_adv, &adv_handle);
+	if (err) {
+		FAIL("Failed to obtain adv handle: %d\n", err);
+	}
+
+	if (bt_adv_lookup_handle(adv_handle) != ext_adv) {
+		FAIL("Adv handle lookup failed");
+	}
```
I think these are better suited for unittests, rather than BSIM tests

rugeGerritsen replied:
Yes, you are right. Do you have a sample unit tests that can be used as a starting point?

Thalley replied:
I think any of the tests in https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bluetooth/host are suitable for that

Thalley commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -131,6 +150,15 @@ int bt_hci_get_adv_handle(const struct bt_le_ext_adv *adv, uint8_t *adv_handle);
+ * @retval The corresponding periodic advertising sync set object on success,
+ *         NULL if it does not exist.
+ */
+struct bt_le_per_adv_sync *bt_hci_get_per_adv_sync(uint16_t handle);
```
I think the `lookup_handle` suffix works fine
```suggestion
struct bt_le_per_adv_sync *bt_hci_per_adv_sync_lookup_handle(uint16_t handle);
```

rugeGerritsen committed:
Bluetooth: HCI: Expose bt_hci_per_adv_sync_lookup_handle()

When implementing vendor specific HCI APIs and events,
we want to be able to convert between host objects,
handles and back again.

Exposing this API makes that possible.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen committed:
Bluetooth: HCI: Expose bt_hci_adv_lookup_handle()

When implementing vendor specific HCI APIs and events,
we want to be able to convert between host objects,
handles and back again.

Exposing this API makes that possible.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>

rugeGerritsen commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -113,6 +113,15 @@ int bt_hci_cmd_send_sync(uint16_t opcode, struct net_buf *buf,
+/** @brief Get connection given a connection handle.
+ *
+ * @param handle The connection handle
+ *
```
TODO: Add note: The returned conn object carries a reference
